libs/docs/src/docs/2.2.0/innerloop-vs-outerloop.md

```diff
@@ -0,0 +1,420 @@
+    - ***Note**: If you are using odov3, you need to create an additional endpoint for the debugging port. See [Odo Spec Support](#odo-spec-support) for more information*
+    - The environment variable definition `DEBUG_PORT` is to set the debug port of the node application
+
+    ```yaml
+    components:
+      - name: runtime
+        container:
+          image: registry.access.redhat.com/ubi8/nodejs-16:latest
+          args: ['tail', '-f', '/dev/null']
+          memoryLimit: 1024Mi
+          mountSources: true
+          env:
+            - name: DEBUG_PORT
+              value: '5858'
+          endpoints:
+            - name: http-node
+              targetPort: 3000
+            - exposure: none
+              name: debug
+              targetPort: 5858
+    ```
+
+3. Now the runtime component has been defined, we need a way of running our project
+    1. Add a command to `npm install` our project packages and mark it as our projects build command
+    2. Add another command to run the main process or the server process `npm start` in this example and mark it as our projects run command
+    3. With these command you should now be able to run a development deployment of your project, for example `odo dev`
+
+    ```yaml
+    commands:
+      - id: install
+        exec:
+          component: runtime
+          commandLine: npm install
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: build
+            isDefault: true
+      - id: run
+        exec:
+          component: runtime
+          commandLine: npm start
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: run
+            isDefault: true
+    ```
+
+4. For testing and debugging we will also need to define commands for each with the shell commands used to debug and test your project
+    - The `kind` of commands will be `debug` and `test` for running your debugging and testing respectively
+
+    ```yaml
+    commands:
+      - id: install
+        exec:
+          component: runtime
+          commandLine: npm install
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: build
+            isDefault: true
+      - id: run
+        exec:
+          component: runtime
+          commandLine: npm start
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: run
+            isDefault: true
+      - id: debug
+        exec:
+          component: runtime
+          commandLine: npm run debug
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: debug
+            isDefault: true
+      - id: test
+        exec:
+          component: runtime
+          commandLine: npm test
+          workingDir: ${PROJECT_SOURCE}
+          group:
+            kind: test
+            isDefault: true
+    ```
+
+```yaml {% title="Final innerloop devfile" filename="devfile.yaml" %}
```

Comment:
add a header above the final devfile snippet

I wonder if it would be better to show the full devfile first and then the procedure

Reply:
@mike-hoang

> add a header above the final devfile snippet

The `title` field adds a header above the code snippet, I could change this to a markdown header if you feel it can be shown a bit better.

> I wonder if it would be better to show the full devfile first and then the procedure

From my own perspective I disagree with this as I feel it would be overwhelming for a reader to see the full source block at the beginning, *however* I think this opens a pretty good discussion to have about what would be a better format for more readable documentation. Currently, the doc pages I've worked with are setup with the full source at the end as what is seen here.

Reply:
I'm ok with keeping the format as is, just wanted to bring it up as I've seen it done both ways before